### ISSUE TITLE
surefire-junit47 reports results against incorrect test

### ISSUE THREAD

**Issue by joedj** [cucumber/cucumber-jvm#865]
I originally raised this issue against the maven-surefire-plugin because the results are different when using the surefire-junit4 provider vs the surefire-junit47 provider, but was instructed to raise it as a cucumber issue: https://issues.apache.org/jira/browse/SUREFIRE-1156

There is a minimal non-working example here:

https://github.com/joedj/surefire-test.git

The project includes two tests - Test1, a simple junit test, and Test2, a cucumber test.  Test2 prints "Test2 Output".

When run with surefire-junit4, both the tests run, and the output is correct:

```
surefire-test$ mvn clean test -Pjunit4 | grep Running
Running Test1
Running Test2
surefire-test$ grep -rl 'Test2 Output' target/surefire-reports
target/surefire-reports/Test2-output.txt
```

When run with surefire-junit47, both the tests run, but the "Running Test2" line is never printed, and the Test2 output appears in the wrong file:

```
surefire-test$ mvn clean test -Pjunit47 | grep Running
Running Test1
surefire-test$ grep -rl 'Test2 Output' target/surefire-reports
target/surefire-reports/Test1-output.txt              # <--- wrong
```

**Comment by joedj** [comment]
Tibor made the following comments on the surefire issue tracker:

    I have seen same issue in JUnit project, but finding those missing among 900 test was postponing the investigation on my side.
    AFAIK there was instantiation problem with Cucumber Runner as talked about it in JUnit group. 
    Isn't any exception in the reports ?

(there is no exception in the reports)

    Cucumber is running Test2 three times and Test1 once.
    Altogether 4.
    The Cucumber changes the description to "Scenario: Scenario:" instead of class name and now it looks like Cucumber does not call testStarted() on org.junit.runner.notification.RunListener.

(I'm not sure that Test2 is actually running three times - I think that's just the odd test-counting behaviour)

    Open bug in Cucumber. Feel free to reopen with new feedback from Cucumber.

**Comment by mpkorstanje** [comment]
Surefire uses [NonConcurrentRunListener#describesNewTestSet](https://github.com/apache/maven-surefire/blob/master/surefire-providers/surefire-junit47/src/main/java/org/apache/maven/surefire/junitcore/NonConcurrentRunListener.java#L120) to determine if a new file needs to be made.

This relies on the implicit class name of the description.

**Comment by mpkorstanje** [comment]
This will be fixed by #1765 but also requires that issues mentioned in [SUREFIRE-1691](https://issues.apache.org/jira/browse/SUREFIRE-1691) are resolved.

**Comment by mpkorstanje** [comment]
As noted in SUREFIRE-1691, this will be fixed eventually and by the looks of it in a way that does not depend on emitting events at feature level.
